packages/broker/src/plugins/operator/ContractFacade.ts

```diff
@@ -0,0 +1,371 @@
+import { Provider } from '@ethersproject/providers'
+import { Operator, Sponsorship, operatorABI, sponsorshipABI } from '@streamr/network-contracts'
+import { StreamID, ensureValidStreamPartitionIndex, toStreamID } from '@streamr/protocol'
+import { EthereumAddress, Logger, TheGraphClient, addManagedEventListener, toEthereumAddress } from '@streamr/utils'
+import { Contract } from 'ethers'
+import sample from 'lodash/sample'
+import fetch from 'node-fetch'
+import { NetworkPeerDescriptor } from 'streamr-client'
+import { OperatorServiceConfig } from './OperatorPlugin'
+
+interface RawResult {
+    operator: null | { latestHeartbeatTimestamp: string | null }
+}
+
+interface EarningsData {
+    sponsorshipAddresses: EthereumAddress[]
+    sumDataWei: bigint
+    maxAllowedEarningsDataWei: bigint
+}
+
+export const VOTE_KICK = '0x0000000000000000000000000000000000000000000000000000000000000001'
+export const VOTE_NO_KICK = '0x0000000000000000000000000000000000000000000000000000000000000000'
+
+export class ParseError extends Error {
+    public readonly reasonText: string
+
+    constructor(reasonText: string) {
+        super(`Failed to parse metadata: ${reasonText}`)
+        this.reasonText = reasonText
+    }
+}
+
+export function parsePartitionFromReviewRequestMetadata(metadataAsString: string | undefined): number | never {
+    if (metadataAsString === undefined) {
+        throw new ParseError('no metadata')
+    }
+
+    let metadata: Record<string, unknown>
+    try {
+        metadata = JSON.parse(metadataAsString)
+    } catch {
+        throw new ParseError('malformed metadata')
+    }
+
+    const partition = Number(metadata.partition)
+    if (isNaN(partition)) {
+        throw new ParseError('invalid or missing "partition" field')
+    }
+
+    try {
+        ensureValidStreamPartitionIndex(partition)
+    } catch {
+        throw new ParseError('invalid partition numbering')
+    }
+
+    return partition
+}
```

Comment:
Of course you'd have to convert that `ZodError` into `ParseError` but this could be a decent entry point into a fruitful adventure of a `toParseError(e: ZodError)` utility. 😉

Reply:
Looks actually pretty nice with `zod`, I'll keep this in mind. But maybe this change can be made later.